File: Napoleon.ce/Core/Napoleon.Net/GRPacket/ByteStream.cs
/*
 * Copyright (C), 2009 - 2010, Гильдия разработчиков
 * 
 * Библиотека для обмена пакетами с сервером
 * ByteStream
 * 
 * ert   14/11/2009   creating
 */
using System;
using System.Collections.Generic;
using System.Text;
using System.Collections;
using System.IO;
using ICSharpCode.SharpZipLib.Checksums;
using ICSharpCode.SharpZipLib.Zip.Compression.Streams;
using System.Threading;
using System.Net.Sockets;

namespace GRSoft.Network
{
   public class ByteStream
   {
      internal static string PACKET_STR_TAG = "GRPACKET";
      internal static string DATA_TAG = "DATA";

      internal static Byte[] PACKET_TAG = Encoding.Unicode.GetBytes(PACKET_STR_TAG);


      int curSym = 0;
      Byte[] bytes;

      public ByteStream(Byte[] bytes)
      {
         this.bytes = bytes;
      }

      public ByteStream(Packet packet)
      {
         CopyBytes(packet);
      }

      public ByteStream(PacketObject po)
      {
         Packet p = new Packet();
         po.ToPacket(p);
         CopyBytes(p);
      }

      private void CopyBytes(Packet packet)
      {
         bytes = new byte[packet.Size];
         int offset = 0;
         foreach (Byte[] pckt in packet)
         {
            int len = pckt.Length;
            Array.Copy(pckt, 0, bytes, offset, len);
            offset += len;
         }
      }

      public byte[] CopyBytes()
      {
         byte[] ret = new byte[bytes.Length];
         Array.Copy(bytes, ret, bytes.Length);
         return ret;
      }

      public bool EOS
      {
         get
         {
            return (bytes == null) ? true : (curSym >= bytes.Length);
         }
      }

      public char Current()
      {
         return GetChar(curSym);
      }

      public char Next()
      {
         return GetChar(curSym + 2);
      }

      char GetChar(int pos)
      {
         if (!EOS)
         {
            if (pos <= bytes.Length - 2)
               return Encoding.Unicode.GetChars(bytes, pos, 2)[0];

            if (pos == bytes.Length - 1)
            {
               Byte[] b = new Byte[2];
               b[0] = bytes[pos];
               b[1] = (byte)0;

               return Encoding.Unicode.GetChars(b)[0];
            }
         }
         return '\0';
      }

      public void MoveToFirst()
      {
         curSym = 0;
      }

      public bool MoveNext()
      {
         if (EOS) return false;

         if (curSym < bytes.Length - 2)
            curSym += 2;
         else
            curSym = bytes.Length;
         return true;
      }

      public bool EatWhite()
      {
         while (!EOS)
         {
            char sym = Current();
            if (!Char.IsWhiteSpace(sym)) return true;

            MoveNext();
         }

         return false;
      }

      public bool CopyUntill(out string res, char stopSym)
      {
         StringBuilder str = new StringBuilder();

         while (!EOS)
         {
            char sym = Current();
            if (sym == stopSym) break;

            str.Append(sym);
            MoveNext();
         }

         res = str.ToString();
         return !EOS;
      }

      internal bool CopyBytes(byte[] dest)
      {
         int sz = dest.Length;
         if (bytes.Length - curSym < sz) return false;

         Array.Copy(bytes, curSym, dest, 0, sz);

         curSym += sz;
         if (sz % 2 != 0) curSym++;

         return true;
      }

      static bool WaitPacket(Stream stream)
      {
         Byte[] buf = new Byte[PACKET_TAG.Length];
         while (true)
         {
            int i;
            int len = stream.Read(buf, 0, buf.Length);
            if (len < buf.Length)
               return false;

            for (i = PACKET_TAG.Length - 1; i >= 0; i--)
               if (buf[i] != PACKET_TAG[i]) break;

            if (i < 0) break;
         }

         return true;
      }

      static bool ReadTill(out string str, Stream stream, char stopSym)
      {
         StringBuilder sb = new StringBuilder();

         Byte[] buf = new Byte[2];

         str = "";
         while (true)
         {
            if (stream.Read(buf, 0, 2) != 2) return false;
            char sym = Encoding.Unicode.GetChars(buf)[0];

            if (sym == stopSym) break;
            sb.Append(sym);
         }

         str = sb.ToString();
         return true;
      }

      internal static bool ReadOptions(List<PacketOption> ops, Stream stream)
      {
         if (!WaitPacket(stream)) return false;

         // get packet options
         while (true)
         {
            string str;
            if (!ReadTill(out str, stream, ';')) return false;

            if (str.CompareTo(DATA_TAG) == 0) break;

            PacketOption op = new PacketOption(str);
            ops.Add(op);
         }

         return (ops.Count != 0);
      }

      public void Send(Stream stream, string operations)
      {
         string head = DATA_TAG + ";";
         if (operations.IndexOf(PacketObject.GZIP_TAG) < 0 && operations.IndexOf(PacketObject.CRC_TAG) < 0)
            operations += PacketObject.CRC_TAG;

         int sp = 0;
         int len = operations.Length;
         while (true)
         {
            int ep = operations.IndexOf(':', sp);
            string op = operations.Substring(sp, (ep < 0) ? len - sp : ep - sp);

            string param;
            bytes = PacketOperator.Encode(bytes, op, out param);
            head = param + head;

            if (ep < 0) break;
            sp = ep + 1;
         }

         string pktTag = PACKET_STR_TAG + "(" + bytes.Length.ToString() + ");";
         head = pktTag + head;

         byte[] headBytes = Encoding.Unicode.GetBytes(head);

         stream.Write(headBytes, 0, headBytes.Length);
         stream.Write(bytes, 0, bytes.Length);
      }

      public static ByteStream Receive(Stream stream, IProgress progress)
      {
         List<PacketOption> ops = new List<PacketOption>();
         if (!ReadOptions(ops, stream)) return null;

         int packetSize = Int32.Parse(ops[0].value);
         ops.RemoveAt(0);

         Byte[] packet = new Byte[packetSize];

         bool setProgress = false;
         if (progress != null && packetSize > 20000)
         {
            progress.SetText(String.Format("Чтение данных {0} КБ", packetSize / 1024));
            progress.SetMax(packetSize);
            setProgress = true;
         }

         int offset = 0;
         NetworkStream ns = stream as NetworkStream;
         while (packetSize > 0)
         {
            int to = ns.ReadTimeout / 10;
            while (!ns.DataAvailable && to > 0)
            {
               Thread.Sleep(10);
               to -= 10;
            }

            int readed = stream.Read(packet, offset, packetSize);
            if (readed == 0)
               break;

            if (setProgress)
               progress.AdvancePos(readed);

            packetSize -= readed;
            offset += readed;
         }

         if (packetSize > 0) return null;

         return DecodePacket(packet, ops);
      }

      private static ByteStream DecodePacket(byte[] packet, List<PacketOption> ops)
      {
         foreach (PacketOption op in ops)
         {
            packet = PacketOperator.Decode(packet, op);
            if (packet == null) break;
         }

         return (packet != null) ? new ByteStream(packet) : null;
      }
   }

   delegate byte[] Decoder(byte[] src, string value);
   delegate byte[] Encoder(byte[] src, out string value);
   class PacketOperator
   {
      struct Operator
      {
         internal string name;
         internal Decoder decoder;
         internal Encoder encoder;


         internal Operator(string n, Decoder decode, Encoder encoder)
         {
            this.name = n;
            this.decoder = decode;
            this.encoder = encoder;
         }
      }

      static Operator[] operators = new Operator[]
      {
         new Operator(PacketObject.GZIP_TAG, Decompress, Compress),
         new Operator(PacketObject.CRC_TAG, CheckCRC32, SetCRC32)
      };

      static byte[] Compress(byte[] src, out string value)
      {
         int len = src.Length;
         if (len < 1024) len = 1024;
         byte[] dest = new byte[len];
         MemoryStream ms = new MemoryStream(dest);

         DeflaterOutputStream compress = new DeflaterOutputStream(ms);
         try
         {
            compress.Write(src, 0, src.Length);
            compress.Finish();
            Array.Resize(ref dest, (int)compress.Position);
         }
         catch
         {
         }

         value = "(" + src.Length.ToString() + ")";
         return dest;
      }

      static byte[] Decompress(byte[] src, string value)
      {
         byte[] dest = new byte[int.Parse(value)];

         MemoryStream ms = new MemoryStream(src);
         Stream decompress = new InflaterInputStream(ms);

         try
         {
            decompress.Read(dest, 0, dest.Length);
         }
         catch
         {
            dest = null;
         }
         return dest;
      }

      static byte[] SetCRC32(byte[] src, out string value)
      {
         Crc32 crc = new Crc32();
         crc.Update(src);
         int crcVal = (int)crc.Value;
         value = "(" + crcVal.ToString() + ")";

         return src;
      }

      static byte[] CheckCRC32(byte[] src, string value)
      {
         Crc32 crc = new Crc32();
         crc.Update(src);
         int c1 = (int)crc.Value;
         int c2 = int.Parse(value);
         if (c1 != c2)
            return null;
         return src;
      }

      internal static byte[] Decode(byte[] src, PacketOption option)
      {
         byte[] dest = null;
         foreach (Operator op in operators)
         {
            if (op.name.CompareTo(option.name) == 0)
            {
               dest = op.decoder(src, option.value);
               break;
            }
         }

         return dest;
      }

      internal static byte[] Encode(byte[] src, string opName, out string param)
      {
         param = "";
         byte[] dest = null;
         foreach (Operator op in operators)
         {
            if (op.name.CompareTo(opName) == 0)
            {
               string val = "";
               dest = op.encoder(src, out val);
               param = opName + val + ";";
               break;
            }
         }

         return dest;
      }
   }

   struct PacketOption
   {
      internal PacketOption(string str)
      {
         int pos = str.IndexOf('(');
         if (pos < 0)
         {
            name = str;
            value = "";
         }
         else
         {
            int ep = str.IndexOf(')', pos + 1);
            if (ep < 0) ep = str.Length - 1;

            name = str.Substring(0, pos);
            value = str.Substring(pos + 1, ep - pos - 1);
         }
      }

      internal string name;
      internal string value;
   }
}

File: Napoleon.ce/Core/Napoleon.AS/dev5/base/src/main/java/com/grsoft/network/ByteStream.java
/*
 * Copyright (C), 2010, Гильдия Разработчиков
 *
 *
 *
 * kki   14/09/2010   creating
 */

package com.grsoft.network;
import static com.grsoft.util.Debug.dbgPrint;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.CRC32;
import java.util.zip.InflaterInputStream;

import com.grsoft.dataobjects.DataObject;
import com.grsoft.napoleon.util.ConvertConstants;
import com.grsoft.network.exception.EndOfStream;
import com.grsoft.network.exception.RuntimeException;
import com.grsoft.network.exception.UnexpectedCharInStream;
import com.grsoft.network.util.UnicodUtils;

import android.content.Context;
import android.util.Log;

public class ByteStream
{
	public static final String PACKET_STR_TAG = "GRPACKET";
	public static final String DATA_TAG = "DATA";
	public static final String GZIP_TAG = "GZIP";
	public static final String CRC_TAG = "CRC";
	
	
	static public int MAX_BUF_LENGTH = 7 * 1024 * 1024;
	
	protected int curSym = 0;
	protected Buffer bytes;
	
	protected int received = 0;
	protected Context context = null;
	
	protected final char END_OF_STRING = '\0';
	
	public ByteStream(byte[] bytes, Context context) {
		this.bytes = new Buffer(bytes);
		this.context = context;
	}
	
	
	
	ByteStream(Context context, int received) {
		this.context = context;
		this.received = received;
	}


	public void close() {}

	public int getReceived() { return received; }
	
	public char current() { return getCharAt(curSym); }

	public char next() { return getCharAt(curSym + 2); }
	
	public static ByteStream createStreamFromDataObjects(List<? extends DataObject> dataObjects ) {
		return null;
	}
	
	private char getCharAt(int pos) {
		return !isEOS()
			? readCharFromBuffer(pos)
			: END_OF_STRING;
	}

	private char readCharFromBuffer(int pos) {
		return (pos <= bytes.getLength() - 2)
			? UnicodUtils.readChar(bytes.getData(), pos, END_OF_STRING)
			: (char)bytes.getData()[pos];
	}

	public boolean moveNext() {
		if (isEOS()) return false;
		
		if (curSym < bytes.getLength() - 2)
			curSym += 2;
		else
			curSym = bytes.getLength();
		
		return true;
	}
	
	public boolean eatWhite() {
		while(!isEOS()) {
			char sym = current();
			
			if(Character.isWhitespace(sym))
				return true;
			
			moveNext();
		}
		
		return false;
	}
	
	public boolean copyUntill(StringBuilder outRes, char stopSym ) {
		while(!isEOS()) {
			char sym = current();
			
			if (sym == stopSym) 
				break;
			
			outRes.append(sym);
			moveNext();
		}
		
		return !isEOS();
	}
	
	public String readStringTillChar(char stopSym) throws RuntimeException {
		StringBuilder result = new StringBuilder();
		
		if (!copyUntill(result, stopSym))
			throw new RuntimeException(new EndOfStream());
		
		return result.toString();
	}
	
	public boolean writeBytes(FileOutputStream fos, int size) throws IOException {
		moveNext();
		if (bytes.getLength() - curSym < size )
			return false;
		
		fos.write(bytes.getData(), curSym, size);

		curSym += size;		
		if(size % 2 != 0) 
			curSym++;
		
		return true;
	}
	
	public boolean copyBytes(byte[] dest) {
		moveNext();
		
		int sz = dest.length;
		
		if (bytes.getLength() - curSym < sz )
			return false;
		
		System.arraycopy(bytes.getData(), curSym, dest, 0, sz);
		
		curSym += sz;
		
		if(sz % 2 != 0) 
			curSym++;
		
		return true;
	}
	
	private static boolean waitPacket(InputStream stream) throws RuntimeException
	{
		byte[] buf = new byte[UnicodUtils.toBytes(PACKET_STR_TAG).length];
		
		while(true)
		{
			int i = 0;
			try
			{
				int len = stream.read(buf, 0, buf.length);
				if(len < buf.length)
					return false;
				
				for(i = UnicodUtils.toBytes(PACKET_STR_TAG).length - 1; i >= 0; i--)
					if (buf[i] != UnicodUtils.toBytes(PACKET_STR_TAG)[i])
						break;
				
				if(i < 0)
					break;
			} catch (IOException e)
			{
				e.printStackTrace();
				return false;
			}
		}
		
		return true;
	}
	
	static boolean readTill(StringBuilder outStr, InputStream stream, char stopSym) {
		byte[] buf = new byte[2];
		
		while(true) {
			try {
				if(stream.read(buf,0,2) != 2)
					return false;
				
				ByteArrayInputStream inputStream = new ByteArrayInputStream(buf);
				InputStreamReader reader = new InputStreamReader(inputStream);
				char sym = (char)reader.read();
				
				if (sym == stopSym)
					break;
				
				outStr.append(sym);
				
			} catch (IOException e) {
				e.printStackTrace();
				return false;
			}
		}
		
		return true;
	}
	
	private static int readOptions(List<PacketOptions> ops, InputStream stream) 
		throws RuntimeException
	{
		int receivedBytes = 0;
		if (!waitPacket(stream))
			return receivedBytes;
		
		while(true)
		{
			StringBuilder sb = new StringBuilder();
			
			if (!readTill(sb, stream, ';')) 
				return 0;
			
			if (sb.toString().equals(DATA_TAG))
				break;	
			
			PacketOptions op = new PacketOptions(sb.toString());
			receivedBytes += (sb.length() * 2);
			ops.add(op);
		}
		
		return (ops.size() != 0) ? receivedBytes : 0;
	}
	
	public boolean isEOS()
	{
		return (bytes == null ) ? true  : curSym >= bytes.getLength();
	}

	public int send(OutputStream stream, String operations) throws RuntimeException
	{
		String head = DATA_TAG + ";";
		if(operations.indexOf(GZIP_TAG) < 0
				&& operations.indexOf(CRC_TAG) < 0)
			operations += CRC_TAG;
		
		int sp = 0;
		int len = operations.length();
		int sendSize = 0;
		
//		try {
//			final String FILE_LOG = "test" + Integer.toString(SrcDataCounter.getValue()) + ".log";
//			File f = new File( Environment.getExternalStorageDirectory().getPath(), FILE_LOG);
//			OutputStream os = new FileOutputStream(f);
//			os.write(bytes.getData());
//			os.close();
//		} catch (Exception e1) {
//			e1.printStackTrace();
//		}
		
		while(true)
		{
			int ep = operations.indexOf(':',sp);
			String op = operations.substring(sp, (ep < 0) ? len - sp : ep - sp);
			
			EncodeResult encodeResult = PacketOperator.encode(bytes.getData(), op);
			String param = encodeResult.getParam();
			bytes = new Buffer(encodeResult.getBuf());
			
			head = param + head;
			
			if (ep < 0 )
				break;
			
			sp = ep + 1;
		}
		
		String pktTag = PACKET_STR_TAG + "(" + Integer.toString(bytes.getLength()) + ");";
		head = pktTag + head;
		
		dbgPrint("head >>> %s", head);
		
//		dbgWriteFileLn(head);
//		dbgWriteFileLn(bytes.getData(), true);
		
		byte[] headBytes = new byte[]{};
		headBytes = UnicodUtils.toBytes(head);
		
		try
		{
			stream.write(headBytes, 0, headBytes.length);
			stream.write(bytes.getData(), 0, bytes.getLength());
			sendSize = headBytes.length + bytes.getLength();
		} catch (IOException e)
		{
			e.printStackTrace();
		}
		
		return sendSize;
	}

	public static ByteStream receive(InputStream stream, Context context) throws RuntimeException {
		ArrayList<PacketOptions> ops = new ArrayList<PacketOptions>();
		
		int rcvd = readOptions(ops, stream);
		if( rcvd == 0 )
			return null;
		
		int packetSize = Integer.parseInt(ops.get(0).getValue());
		ops.remove(0);
		
		rcvd += packetSize;
		byte[] packet = readFully(stream, packetSize);
//		ByteStream res = decodePacket(packet, ops);

		ByteStream inStr = new ByteStream(packet, context);
		ByteStream res = decodePacket(inStr, ops);
		if( res != null )
			res.received = rcvd;
		
		return res;
	}
	
	private static byte[] readFully(InputStream stream, int size) throws RuntimeException {
		try {
			byte[] result = new byte[size];
			
			DataInputStream dataInputStream = new DataInputStream(stream);
			dataInputStream.readFully(result);
		
			return result;
		}
		catch(Exception exception) {
			throw new RuntimeException(exception);
		}
		
	}
	
	ByteStream decompress(String value) {
		Runtime r = Runtime.getRuntime();
		long maxMem = r.maxMemory();
		long freeMem = r.freeMemory();
		long avail = 2 * (maxMem - (r.totalMemory() - freeMem)) / 3;

		final int BUF_SIZE = Integer.parseInt(value);
//		if(BUF_SIZE > MAX_BUF_LENGTH)
		if(BUF_SIZE > avail)
			return FileByteStream.decompress(bytes.getData(), received, context);
		
		byte[] dest = new byte[BUF_SIZE];
		
		ByteArrayInputStream ms = new ByteArrayInputStream(bytes.getData());
		InflaterInputStream decompress = new InflaterInputStream(ms);
		
		try {
			int offset = 0;
			int bytes_read = 0;
			
			while(offset < BUF_SIZE) {
				bytes_read = decompress.read(dest, offset, BUF_SIZE - offset);
				offset += bytes_read;
			}
			
		} catch (IOException e) {
			e.printStackTrace();
			close();
			return null;
		}
		
		close();
		return new ByteStream(dest, context);
	}
	
	ByteStream checkCRC32(String value) {
		int c2 = Integer.parseInt(value);
		CRC32 crc = new CRC32();
		crc.update(bytes.getData());
		int c1 = (int)crc.getValue();
		if(c1 != c2)
			return null;
		return this;
	}

	private static ByteStream decodePacket(ByteStream packet, ArrayList<PacketOptions> ops) {
		for (PacketOptions op : ops) {
			String opName = op.getName();
			
			if(ByteStream.GZIP_TAG.equals(opName)) {
				packet = packet.decompress(op.getValue());
			} else if(ByteStream.CRC_TAG.equals(opName)) {
				packet = packet.checkCRC32(op.getValue());
			} else {
				packet = null;
			}
				

			if (packet == null)
				break;
		}
		
		return packet;
	}
	

//	private static ByteStream decodePacket(byte[] packet, ArrayList<PacketOptions> ops)
//	{
//		for (PacketOptions op : ops)
//		{
//			packet = PacketOperator.decode(packet, op);
//			if (packet == null)
//				break;
//		}
//		
//		/*------------------------------ DEBUG SECTION ---------------------------*/
////		try{
////			File f = new File(Environment.getExternalStorageDirectory(), "receive.txt");
////			FileOutputStream fos = new FileOutputStream(f);
////			fos.write(packet);
////			fos.close();
////		}catch(Exception e){
////			e.printStackTrace();
////		}
//		/*------------------------------ DEBUG SECTION ---------------------------*/
//		
//		return (packet != null) ? new ByteStream(packet) : null;
//	}
	
	@Override
	public String toString()
	{
		try
		{
			return new String(bytes.getData(), "UTF_16LE");
		}
		catch(Exception exception)
		{
			return new String("ByteStream can't representation as String" + exception.getMessage());
		}	
	}
	
	public String readString() throws UnexpectedCharInStream
	{
		ByteStreamReader byteStreamReader = new ByteStreamReader(this);
		return byteStreamReader.readString();
	}
	
	public int getSize()
	{
		return bytes.getLength();
	}
	
	public int getPosition()
	{
		return curSym;
	}
	
//	public boolean isContinues(){
//		final byte[] CONTINUE_FLAG = new byte[]{83, 0, 116, 0, 114, 0, 101, 0, 97, 0, 109, 0, 
//				67, 0, 111, 0, 110, 0, 116, 0, 105, 0, 110, 0, 117, 0, 101, 0, 91, 0, 99, 0, 
//				111, 0, 110, 0, 116, 0, 105, 0, 110, 0, 117, 0, 101, 0, 58, 0, 110, 0, 93, 0, 
//				91, 0, 49, 0, 93, 0};
//		final int CONTINUE_FLAG_LENGH = CONTINUE_FLAG.length;
//		
//		try{
//			if (bytes.getLength() > CONTINUE_FLAG_LENGH){
//				byte[] test = new byte[CONTINUE_FLAG_LENGH];
//				
//				System.arraycopy(bytes.getData(), bytes.getLength() - CONTINUE_FLAG_LENGH, 
//						test, 0, CONTINUE_FLAG_LENGH);
//				
//				return Arrays.equals(test, CONTINUE_FLAG);
//			}
//			
//			return false;
//		}catch(Exception e){
//			return false;
//		}
//	}
}

class Buffer
{
	private byte[] data;
	private int length; 
	
	public Buffer(byte[] data)
	{
		this.data = data;
		this.length = data.length;
	}
	
	public byte[] getData()
	{
		return data;
	}
	
	public int getLength()
	{
		return length;
	}
}

class ByteStreamReader
{
	ByteStream context;
	
	public ByteStreamReader(ByteStream context)
	{
		this.context = context;
	}
	
	public String readString() throws UnexpectedCharInStream
	{
		if (context.current() != ConvertConstants.QUOTE)
	           throw new UnexpectedCharInStream(context);

	    context.moveNext();

	    String result = readBodyString();
	        
	    context.moveNext();
	        
	    return result.toString();
	}
	
	private String readBodyString() throws UnexpectedCharInStream
	{
		StringBuilder result = new StringBuilder(32);
		
		while (context.current() != ConvertConstants.QUOTE)
        {
        	if (context.isEOS())
        	   throw new UnexpectedCharInStream(context);
           
        	char sym = context.current();
        	
        	if (sym == ConvertConstants.SLASH)
        	{
        		if (!context.moveNext())
        			throw new UnexpectedCharInStream(context);
        		
        		sym = convertStreamChar();
        	}
          
        	result.append(sym);
        	context.moveNext();
        }
		
		return result.toString();
	}
	
	private char convertStreamChar() throws UnexpectedCharInStream
	{
		switch (context.current())
        {
			case '\\': 
        	case '/': 
        	case '"': 
        	   return context.current();
        	   
        	case 'b': 
        	   return '\b';
        	   
        	case 'f': 
        	   return '\f';
        	   
        	case 'n': 
        	   return '\n';
        	   
        	case 'r': 
        	   return '\r'; 
        	   
        	case 't': 
        	   return '\t';
        	
        	default:
        	   throw new UnexpectedCharInStream(context);
         }
	}
	
	
}
